### ISSUE TITLE
Specific need for PlatRock usage

### ISSUE THREAD

**Issue by batzkass** [siconos/siconos#409]
Hi,

I would like to implement the forest into PlatRock 3D propagation software (Python) with Siconos engine. For the record, the current code (without forest) is hosted [here](https://gitlab.com/platrock/platrock/-/blob/master/platrock/ThreeD/Engines.py#L263), and the main computation loop is located [here](https://gitlab.com/platrock/platrock/-/blob/master/platrock/ThreeD/Engines.py#L349). I use MechanicsIO with my own set of commands to setup and run siconos computations. Terrain is generated with `SiconosHeightMap` and rocks Rocks with  `SiconosBox`. All other details should be easily readable from above links.

Once detected, the contact computation will be done in Python: we have a whole set of data coming from detailed DEM simulations and forming an abacus. This set of data is used to compute the rock output velocity as a function of its input velocity and the rock-tree relative position. Note that some approximations will be done for this kind of contact, and the most important one is that the rock is seen as its equivalent sphere shape, but we can live with that.

**So my question is about contact detection**. The simplest - but probably slowest - method would be to detect rock-tree contacts from Python roughly, at each Siconos loop. But I would prefer to let siconos/bullet do the job. The way I think this could be achieved : 
1. Add the trees into the Siconos scene as vertical cylinders, with corresponding contactors ;
2. Let siconos detect contacts ;
3. Avoid siconos to handle the contacts, the trees should not have any influence on the rock trajectory from siconos point of view;
4. Retrieve rock-trees contacts into Python at each step, and modify the rocks velocity according to our abacus.

For points 1-3 I have the feeling that siconos is flexible enough to do all I want out of the box, but I'm not sure about the proper way to construct it in Python, could you please give me some advises ?
For point 4, I don't know how to retrieve directly from siconos a list of rock-tree contact(s) at each step, again your help would be helpful.

Many thanks,
François Kneib

**Comment by vacary** [comment]
Hi,
It should be possible to retrieve of detected contact points between the rock and the cylinders after the update of the interactions (contact points) by Bullet, and then to treat them separately. For that, I think that the best way is to explode the `computeOneStep()` method as it is done in `mechanics_run` and then to apply your strategy after the update of Interactions.

**Comment by vacary** [comment]
The exploded version of the Newton step is here https://github.com/siconos/siconos/blob/master/io/swig/io/mechanics_run.py

The update of the interaction is made in s.initialize but we can still further explode this function

```
def explode_Newton_solve(self, with_timer):
        s = self._simulation
        self.log(s.initialize, with_timer)()
        self.log(s.resetLambdas, with_timer)()
        # Again the access to s._newtonTolerance generates a segfault due to director
        #newtonTolerance = s.newtonTolerance()
        newtonMaxIteration = s.newtonMaxIteration()

        # return _kernel.TimeStepping_newtonSolve(self, criterion, maxStep)
        # RuntimeError: accessing protected member newtonSolve
        #s.newtonSolve(newtonTolerance, newtonMaxIteration);

        newtonNbIterations = 0
        isNewtonConverge = False
        explode_computeOneStep = False

        self.log(s.initializeNewtonLoop, with_timer)()
        while (not isNewtonConverge) and newtonNbIterations < newtonMaxIteration:
            #self.print_verbose('newtonNbIterations',newtonNbIterations)
            info = 0
            newtonNbIterations = newtonNbIterations + 1
            self.log(s.prepareNewtonIteration, with_timer)()
            self.log(s.computeFreeState, with_timer)()
            if s.numberOfOSNSProblems() > 0:
                if explode_computeOneStep:
                    # experimental
                    osnsp = s.oneStepNSProblem(SICONOS_OSNSP_TS_VELOCITY)
                    #info = self.log(osnsp.compute, with_timer)(s.nextTime())
                    fc = cast_FrictionContact(osnsp)
                    #self.log(fc.updateInteractionBlocks, with_timer)()
                    self.log(fc.preCompute, with_timer)(s.nextTime())
                    self.log(fc.updateMu, with_timer)()
                    info = self.log(fc.solve, with_timer)()
                    self.log(fc.postCompute, with_timer)()
                else:
                    info = self.log(s.computeOneStepNSProblem, with_timer)(SICONOS_OSNSP_TS_VELOCITY)
            self.log(s.DefaultCheckSolverOutput, with_timer)(info)
            self.log(s.updateInput, with_timer)()
            self.log(s.updateState, with_timer)()
            if (not isNewtonConverge) and (newtonNbIterations < newtonMaxIteration):
                self.log(s.updateOutput, with_timer)()
            isNewtonConverge = self.log(s.newtonCheckConvergence, with_timer)
            if s.displayNewtonConvergence():
                s.displayNewtonConvergenceInTheLoop()
            if (not isNewtonConverge) and (not info):
                if s.numberOfOSNSProblems() > 0:
                    self.log(s.saveYandLambdaInOldVariables, with_timer)()
        if s.displayNewtonConvergence():
            s.displayNewtonConvergenceAtTheEnd(info, newtonMaxIteration)
```

**Comment by vacary** [comment]
In this branch that should be merge in the next weeks, https://github.com/vacary/siconos/blob/optimize_2d_granular/io/swig/io/mechanics_run.py
the Newton step is further exploded and the update of interactions is explicitly shown. 
If you need any further help, we can arrange to draft a prototype with you

**Comment by batzkass** [comment]
Many thanks for the work, and sorry for my late answer. I have some time now to work on it again. However my knowledge of siconos and the underlying numerical methods are limited, so I can't work on my own on this. At first I imagined that it could be possible to set a mask (or kind of) to avoid the physical interactions between two group of bodies, but still detecting their geometrical intersection.
My understanding of your proposed solution is that:
* this kind of mask doesn't exist,
* so I shoud remove the interaction from a list, just after having stored the relevant data in PlatRock. All that should be made in python side, do all the corresponding methods exists, and are they binded to python ?

**Comment by vacary** [comment]
This issue should be fixed with #422
